File: leetcode-daily/march-2024/713-subarray-product-less-than-k.cpp
/*
https://leetcode.com/problems/subarray-product-less-than-k

Given an array of integers `nums` and an integer `k`, return *the number of contiguous subarrays
where the product of all the elements in the subarray is strictly less than *`k`.


**Example 1:**

**Input:** nums = [10,5,2,6], k = 100
**Output:** 8
**Explanation:** The 8 subarrays that have product less than 100 are:
[10], [5], [2], [6], [10, 5], [5, 2], [2, 6], [5, 2, 6]
Note that [10, 5, 2] is not included as the product of 100 is not strictly less than k.

**Example 2:**

**Input:** nums = [1,2,3], k = 0
**Output:** 0


**Constraints:**

* `1 <= nums.length <= 3 * 10⁴`
* `1 <= nums[i] <= 1000`
* `0 <= k <= 10⁶`
*/

#include "./../../cpp-utils/printutils.h"
#include "bits/stdc++.h"
#include <iostream>
using namespace std;

class Solution {
  public:
    // O(n^2) sliding window solution that works
    int numSubarrayProductLessThanK2(vector<int> &nums, int k) {
        if (k == 0)
            return 0;

        int res = 0;
        int product;
        for (int i = 0; i < nums.size(); ++i) {
            product = 1;
            for (int j = i; j < nums.size(); ++j) {
                product *= nums[j];
                if (product >= k)
                    break;
                else
                    res++;
            }
        }

        return res;
    }

    // O(n) solution after looking up some solutions
    // and Use the intuition that if there is a n sized array which is valid
    // then there are close to n^2 subarrays [n + n-1 + n-2 + n-3 + ...].
    // but if we add one more element that will still keep the array valid,
    // then there are n more subarrays due to this element.
    int numSubarrayProductLessThanK(vector<int> &nums, int k) {
        if (k == 0)
            return 0;

        int n = nums.size(), res = 0, product = 1, l = 0, r = 0;
        while (r < n) {
            product *= nums[r];

			// move left pointer until product < k
            while (l <= r && product >= k) {
                product /= nums[l];
                l++;
            }

            res += r - l + 1;
            r++;
        }

        return res;
    }
};

int main() {
    Solution sol;

    vector<int> nums_1 = {10, 5, 2, 6};
    int k_1 = 100;
    int expected_1 = 8;
    int output_1 = sol.numSubarrayProductLessThanK(nums_1, k_1);
    cout << output_1 << endl;
    assert(output_1 == expected_1);

    vector<int> nums_2 = {1, 2, 3};
    int k_2 = 0;
    int expected_2 = 0;
    int output_2 = sol.numSubarrayProductLessThanK(nums_2, k_2);
    cout << output_2 << endl;
    assert(output_2 == expected_2);

    return 0;
}

File: leetcode-daily/march-2024/713-subarray-product-less-than-k.rs
/*
https://leetcode.com/problems/subarray-product-less-than-k

Given an array of integers `nums` and an integer `k`, return *the number of contiguous subarrays where the product of
all the elements in the subarray is strictly less than *`k`.


**Example 1:**

**Input:** nums = [10,5,2,6], k = 100
**Output:** 8
**Explanation:** The 8 subarrays that have product less than 100 are:
[10], [5], [2], [6], [10, 5], [5, 2], [2, 6], [5, 2, 6]
Note that [10, 5, 2] is not included as the product of 100 is not strictly less than k.

**Example 2:**

**Input:** nums = [1,2,3], k = 0
**Output:** 0


**Constraints:**

* `1 <= nums.length <= 3 * 10⁴`
* `1 <= nums[i] <= 1000`
* `0 <= k <= 10⁶`
*/

struct Solution;
impl Solution {
    pub fn num_subarray_product_less_than_k(nums: Vec<i32>, k: i32) -> i32 {
        if k <= 0 {
            return 0;
        }

        let mut res: i32 = 0;
        let mut product: i32 = 1;
        let mut l: usize = 0;
        let mut r: usize = 0;
        let n: usize = nums.len();

        while r < n {
            product *= nums[r];

            while l <= r && product >= k {
                product /= nums[l];
                l += 1;
            }

            res += (r - l + 1) as i32;
            r += 1;
        }

        res
    }
}

fn main() {
    let nums_1: Vec<i32> = vec![10, 5, 2, 6];
    let k_1: i32 = 100;
    let expected_1: i32 = 8;
    let output_1: i32 = Solution::num_subarray_product_less_than_k(nums_1, k_1);
    println!("{:?}", output_1);
    assert!(output_1 == expected_1);

    let nums_2: Vec<i32> = vec![1, 2, 3];
    let k_2: i32 = 0;
    let expected_2: i32 = 0;
    let output_2: i32 = Solution::num_subarray_product_less_than_k(nums_2, k_2);
    println!("{:?}", output_2);
    assert!(output_2 == expected_2);
}
